pkg/api/types.go

```diff
@@ -1410,6 +1414,54 @@ type ReplicationControllerList struct {
 	Items []ReplicationController `json:"items"`
 }
 
+// DisruptionBudget is an object to define the max disruption that can be caused to a collection of pods
+type DisruptionBudget struct {
+	// Specification of the desired behavior of the DisruptionBudget.
+	Spec DisruptionBudgetSpec `json:"spec,omitempty"`
+	// Most recently observed status of the DisruptionBudget.
+	Status DisruptionBudgetStatus `json:"status,omitempty"`
+}
+
+// DisruptionBudgetSpec is a description of a DisruptionBudget.
+type DisruptionBudgetSpec struct {
+	// RateLimits is a list of rate limits on disruptions (preemption and other evictions) across the corresponding
+	// collection of pods, e.g. no more than one disruption per hour across the pods of a particular Service
+	RateLimits []RateLimitSpec `json:"rateLimits,omitempty"`
+	// The minium number of pods that must be available simultaneously.
+	// Value can be an absolute number (ex: 5) or a percentage of total pods(ex: 10%).
+	// Absolute number is calculated from percentage by rounding up.
+	// By default, a fixed value of 1 is used.
+	MinAvailable intstr.IntOrString `json:"minAvailable,omitempty"`
+	// Minimum number of seconds for which a newly created pod should be ready
+	// without any of its container crashing, for it to be considered available.
+	// Defaults to 0 (pod will be considered available as soon as it is ready)
+	MinReadySeconds int `json:"minReadySeconds,omitempty"`
+}
+
+// DisruptionBudgetStatus represents information about the status of a DisruptionBudget. Status may trail the actual
+// state of a system.
+type DisruptionBudgetStatus struct {
+	// The current disruption rate that the collection of pods is experiencing.
+	// This is an array of DeletionTimestamp associated with DisruptionToken: insertion timestamp + RateLimitSpec.Duration.
+	// This data may be out of date by some window of time.
+	DisruptionInfo []unversioned.Time `json:"disruptionInfo,omitempty"`
+	// Total number of available pods (ready for at least minReadySeconds) targeted by this disruption budget.
+	AvailableReplicas int `json:"availableReplicas,omitempty"`
+	// Total number of unavailable pods targeted by this disruption budget.
+	UnavailableReplicas int `json:"unavailableReplicas,omitempty"`
+}
+
+// RateLimitSpec describes the rate limit on disruptions (preemption and other evictions) across the corresponding
+// collection of pods
+type RateLimitSpec struct {
+	// BucketSize is the maximum number of DisruptionToken could exist during a period of time.
+	// Once there are more than BucketSize number of DisruptionToken during a period of time, we can
+	// not disrupt pods in a further step.
+	BucketSize inf.Dec `json:"bucketSize,omitempty"`
+	// Duration defines how long a token lives in the bucket, from insertion until deletion
```

Comment:
The bucket thing is an implementation detail. Just say "the time period over which NumAllowedDisruptionsPerDuration applies."